File: go/src/213.打家劫舍-ii.go
/*
 * @lc app=leetcode.cn id=213 lang=golang
 *
 * [213] 打家劫舍 II
 */

// @lc code=start
func rob(nums []int) int {
	n := len(nums)
	if n == 0 {
		return 0
	}
	if n < 4 {
		return max(nums...)
	} 
	return max(helper(nums, 0, n - 2), helper(nums, 1, n - 1))
}

func helper(nums []int, start, end int) int {
	a, b := nums[start], 0 // a：偷 b：不偷
	for i := start + 1; i <= end; i++ {
		a, b = b + nums[i], max(a, b)
	}
	return max(a, b)
}

func max(a ...int) int {
	res := a[0]
	for _, v := range a {
		if res < v {
			res = v
		}
	}
	return res
}
// @lc code=end

File: rust/src/213.打家劫舍-ii.rs
/*
 * @lc app=leetcode.cn id=213 lang=rust
 *
 * [213] 打家劫舍 II
 */

// @lc code=start
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.len() == 1 {
            return nums[0];
        }

        return Self::util(nums[1..].to_vec()).max(Self::util(nums[..nums.len() - 1].to_vec()));
    }

    fn util(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        if n == 0 {
            return 0;
        }

        let mut dp = vec![0; n + 1];
        dp[1] = nums[0];

        nums.iter().enumerate().skip(1).for_each(|(i, x)| {
            dp[i + 1] = dp[i].max(dp[i - 1] + x);
        });

        dp[n]
    }
}
// @lc code=end
